### PR TITLE
Update common block in sync after importing blocks of a peer, please read UPDATE

### PR DESCRIPTION
**Pull request by bkchr**
This updates the sync code to update the common block of a peer, after
we have imported blocks from this peer. This fixes a bug for when we are
connected to one or more nodes that are doing a full sync as our node.
Nodes in full sync will not announce new blocks, as we don't send import
notifications on full sync. The problem as now that we were connected to
some peer that reported some low number as its best and we tried to sync
these blocks. But, as we did not update the common block of this peer,
we would sync these blocks over and over again. Being captured in some
time warp.
The solution to this problem is that we increase the common number as we
import blocks from this peer.

UPDATE:

Updating the common block didn't fix the entire story :( I added another regression test that is describing the scenario we hit. 

TLDR: We are doing a full resync while being connected to a node that is doing the same. The other node finishes before us and we get a block announcement from its importing something at the tip of the chain. The problem is that our common number didn't change in the sync process, so we would start to request blocks again at our common number (if we are unlucky, this means from genesis..). To prevent requesting the full chain another time, we instead do an ancestor search to find the real common block.

### CODE REVIEW

File: client/network/test/src/sync.rs [hunk 1]
```diff
@@ -814,3 +811,32 @@ fn sync_to_tip_requires_that_sync_protocol_is_informed_about_best_block() {
 	// However peer 1 should still not have the block.
 	assert!(!net.peer(1).has_block(&block_hash));
 }
+
+#[test]
+fn lol() {
```
Comment (kianenigma) [inline client/network/test/src/sync.rs:816]: 🤪

Comment (bkchr) [reply client/network/test/src/sync.rs:816]: Fuck xD Like last time xD

**Review by tomaka — APPROVED** [review]
Looks good to me.
(you have a failing test, though)

File: client/network/src/protocol/sync.rs [hunk 2]
```diff
@@ -2073,4 +2136,69 @@ mod test {
 		// Nothing to import
 		assert!(matches!(res, OnBlockData::Import(_, blocks) if blocks.is_empty()));
 	}
+
+	#[test]
+	fn lol() {
```
Comment (seunlanlege) [inline client/network/src/protocol/sync.rs:2141]: ha lol, should probably change this

File: client/network/src/protocol/sync.rs [hunk 3]
```diff
@@ -687,7 +718,21 @@ impl<B: BlockT> ChainSync<B> {
 				return None
 			}
 
-			if let Some((range, req)) = peer_block_request(
+			// If our best queued is more than `MAX_BLOCKS_TO_LOOK_BACKWARDS` blocks away from the
+			// common number, the peer best number is higher than our best queued and the common
+			// number is smaller than the last finalized block number, we should do an ancestor
+			// search to find a better common block.
+			if best_queued.saturating_sub(peer.common_number) > MAX_BLOCKS_TO_LOOK_BACKWARDS.into()
+				&& best_queued < peer.best_number && peer.common_number < last_finalized
```
Comment (andresilva) [inline client/network/src/protocol/sync.rs:726]: Why do we need the `peer.common_number < last_finalized` condition?

File: client/network/src/protocol/sync.rs [hunk 4]
```diff
@@ -1485,13 +1583,13 @@ pub enum AncestorSearchState<B: BlockT> {
 fn handle_ancestor_search_state<B: BlockT>(
 	state: &AncestorSearchState<B>,
 	curr_block_num: NumberFor<B>,
-	block_hash_match: bool
+	block_hash_match: bool,
 ) -> Option<(AncestorSearchState<B>, NumberFor<B>)> {
 	let two = <NumberFor<B>>::one() + <NumberFor<B>>::one();
-	match state {
+	match dbg!(state) {
```
Comment (andresilva) [inline client/network/src/protocol/sync.rs:1589]:
```suggestion
	match state {
```

File: client/network/src/protocol/sync.rs [hunk 5]
```diff
@@ -1485,13 +1583,13 @@ pub enum AncestorSearchState<B: BlockT> {
...
 		AncestorSearchState::ExponentialBackoff(next_distance_to_tip) => {
-			let next_distance_to_tip = *next_distance_to_tip;
-			if block_hash_match && next_distance_to_tip == One::one() {
+			let next_distance_to_tip = dbg!(*next_distance_to_tip);
```
Comment (andresilva) [inline client/network/src/protocol/sync.rs:1591]:
```suggestion
			let next_distance_to_tip = *next_distance_to_tip;
```

File: client/network/src/protocol/sync.rs [hunk 6]
```diff
@@ -1485,13 +1583,13 @@ pub enum AncestorSearchState<B: BlockT> {
...
+			if dbg!(block_hash_match) && next_distance_to_tip == One::one() {
```
Comment (andresilva) [inline client/network/src/protocol/sync.rs:1592]:
```suggestion
			if block_hash_match && next_distance_to_tip == One::one() {
```

File: client/network/test/src/sync.rs [hunk 7]
```diff
@@ -814,3 +811,32 @@ fn sync_to_tip_requires_that_sync_protocol_is_informed_about_best_block() {
 	// However peer 1 should still not have the block.
 	assert!(!net.peer(1).has_block(&block_hash));
 }
+
+#[test]
+fn make_sure_we_sync_to_tip_without_informing_sync_about_new_best_block() {
```
Comment (andresilva) [inline client/network/test/src/sync.rs:816]: The name of this test conflicts with the name of the test above 🙈

File: client/network/test/src/sync.rs [hunk 8]
```diff
@@ -814,3 +811,32 @@ fn sync_to_tip_requires_that_sync_protocol_is_informed_about_best_block() {
...
+		5_000,
+		false,
+	);
+
+	net.block_until_connected();
+	net.block_until_idle();
+
+	assert!(!net.peer(2).has_block(&block_hash));
+
+	net.peer(0).network_service().new_best_block_imported(block_hash, 10_000);
```
Comment (andresilva) [inline client/network/test/src/sync.rs:838]: Aren't we informing sync about new best block here?

File: client/network/src/protocol/sync.rs [hunk 9]
```diff
@@ -2073,4 +2167,118 @@ mod test {
...
+				res,
+				OnBlockData::Import(_, blocks) if blocks.is_empty()
+			),
+		);
+
+		let peer1_from = unwrap_from_block_number(peer1_req.unwrap().from);
+
+		// As we are on the same chain, we should directly continue with requesting blocks from
+		// peer 2 as well.
+		get_block_request(
```
Comment (andresilva) [inline client/network/src/protocol/sync.rs:2276]: Should we assert that we get some non-empty request here?

**Review by andresilva — CHANGES_REQUESTED** [review]

File: client/network/src/protocol/sync.rs (hunk 9 above)
Comment (bkchr) [reply client/network/src/protocol/sync.rs:2276]: The function is already checking this.

File: client/network/test/src/sync.rs (hunk 7 above)
Comment (bkchr) [reply client/network/test/src/sync.rs:816]: Hmm 😂

File: client/network/src/protocol/sync.rs [hunk 10]
```diff
@@ -67,6 +67,10 @@ const MAX_IMPORTING_BLOCKS: usize = 2048;
 /// Maximum blocks to download ahead of any gap.
 const MAX_DOWNLOAD_AHEAD: u32 = 2048;
 
+/// Maximum blocks to look backwards. The gap is the difference between the highest block and the
+/// common block of a node.
+const MAX_BLOCKS_TO_LOOK_BACKWARDS: u32 = MAX_DOWNLOAD_AHEAD / 2;
```
Comment (arkpar) [inline client/network/src/protocol/sync.rs:72]: Does this mean that the node won't be able to recover from a fork longer than 1024 blocks?

Comment (bkchr) [reply client/network/src/protocol/sync.rs:72]: I added a check that the common block needs to be less than the last finalized block to trigger this behavior.

I can add a test.

File: client/network/src/protocol/sync.rs [hunk 11]
```diff
@@ -1316,18 +1412,19 @@ impl<B: BlockT> ChainSync<B> {
 		}
 
 		if let PeerSyncState::AncestorSearch {..} = peer.state {
+			trace!(target: "sync", "Peer state is ancestor search.");
```
Comment (arkpar) [inline client/network/src/protocol/sync.rs:1415]: This message could use more context, as it may be interleaved with other messages in the log. Perhaps would be better to additionally print `peer.state` [here](https://github.com/paritytech/substrate/pull/7733/files#diff-a5dfc701556975f194fd5d7df78d855b7d2007e741b565025022b849d143c4eeR1372) instead

Comment (bkchr) [reply client/network/src/protocol/sync.rs:1438]: At the top of the function we already print the peer id and nothing else can be printed in between from syncing that isn't related to this peer. So, we don't need to print anymore here.

File: client/network/src/protocol/sync.rs (hunk 3 above)
Comment (bkchr) [reply client/network/src/protocol/sync.rs:738]: If the common number is below the finalized number, the common number must be at least the last finalized.

File: client/network/src/protocol/sync.rs [hunk 12]
```diff
@@ -2073,4 +2206,253 @@ mod test {
...
+	/// A test that ensures that we can sync a huge fork.
+	///
+	/// The following scenario:
+	/// A peer connects to us and we both have the common block 512. The last finalized is 2048.
+	/// Our best block is 4096. The peer send us a block announcement with 4097 from a fork.
+	///
+	/// We will first do an ancestor search to find the common block. After that we start to sync
+	/// the fork and finish it ;)
+	#[test]
+	fn can_sync_huge_fork() {
```
Comment (bkchr) [inline client/network/src/protocol/sync.rs:2332]: @arkpar added a test here for syncing a huge fork.

File: client/network/test/src/sync.rs (hunk 8 above)
Comment (tomaka) [reply client/network/test/src/sync.rs:840]: I think it's intentional. You *have* to inform the sync at some point for the thing to work.

**Review by tomaka — APPROVED** [review]
Starting the ancestry search is more or less the change that I tried doing two days ago, except I was starting the ancestry search immediately when receiving a block announce. But I think your solution is better.

File: client/network/src/protocol/sync.rs (hunk 3 above)
Comment (andresilva) [reply client/network/src/protocol/sync.rs:738]: Got it. This is valid since we are guaranteeing that peer's best number is higher than our best (or queued). But what about chains with no finality where `last_finalized` will always be 0?

File: client/network/src/protocol.rs [hunk 13]
```diff
@@ -528,13 +546,21 @@ impl<B: BlockT, H: ExHashT> Protocol<B, H> {
 
 	/// Inform sync about new best imported block.
 	pub fn new_best_block_imported(&mut self, hash: B::Hash, number: NumberFor<B>) {
+		trace!(target: "sync", "New best block imported {:?}/{}", hash, number);
```
Comment (andresilva) [inline client/network/src/protocol.rs:549]:
```suggestion
		trace!(target: "sync", "New best block imported {:?}/#{}", hash, number);
```

File: client/network/test/src/sync.rs (hunk 8 above)
Comment (andresilva) [reply client/network/test/src/sync.rs:840]: My comment was because the test was named `make_sure_we_sync_to_tip_without_informing_sync_about_new_best_block`, it's been renamed since :)

**Review by andresilva — APPROVED** [review]

File: client/network/src/protocol/sync.rs (hunk 3 above)
Comment (bkchr) [reply client/network/src/protocol/sync.rs:738]: You will run into other problems :see_no_evil: I added some todos for myself about further bugs, however I don't wanted to go deeper into this rabbit hole here :see_no_evil:
